src/github.com/getlantern/lantern-android/client/config.go

```diff
@@ -0,0 +1,122 @@
+package client
+
+import (
+	"compress/gzip"
+	"crypto/x509"
+	"errors"
+	"github.com/getlantern/fronted"
+	"github.com/getlantern/keyman"
+	"github.com/getlantern/yaml"
+	"io"
+	"io/ioutil"
+	"net/http"
+)
+
+type clientCfg struct {
+	FrontedServers []frontedServer                  `yaml:"frontedservers"`
+	MasqueradeSets map[string][]*fronted.Masquerade `yaml:"masqueradesets"`
+}
+
+// Config provides client configuration.
+type Config struct {
+	Client     clientCfg `yaml:"client"`
+	TrustedCAs []*CA     `yaml:"trustedcas"`
+}
+
+var (
+	// ErrFailedConfigRequest is returned when the server replies with a non-200
+	// status code to our request for a configuration file.
+	ErrFailedConfigRequest = errors.New(`Could not get configuration file.`)
+
+	// ErrInvalidConfiguration is returned in case the configuration file is
+	// downloaded but has no useful data.
+	ErrInvalidConfiguration = errors.New(`Invalid configuration file.`)
+)
+
+const (
+	// URL of the configuration file. Remember to use HTTPs.
+	remoteConfigURL = `https://s3.amazonaws.com/lantern_config/cloud.1.6.0.yaml.gz`
+)
+
+// pullConfigFile attempts to retrieve a configuration file over the network,
+// then it decompresses it and returns the file's raw bytes.
+func pullConfigFile() ([]byte, error) {
+	var err error
+	var res *http.Response
+
+	// Issuing a post request to download configuration file.
+	if res, err = http.Get(remoteConfigURL); err != nil {
```

Comment:
@oxtoacart Yeah, you're right, we can't suppose we're going to have this URL always available.

> Immediately start proxying based on the built-in configuration
> Check for the cloud configuration through the already running proxy

This sounds good, I could give it another try.

> Save the configuration to disk for whenever it next starts up (it looks like you're not saving it anywhere?)

We have a limitation on Android and is that writing files is not that easy unless we have a writable path, as far as I know, this path is different for each application namespace, and we don't have that information unless the app provides it to us. We could export a new function that the app could use to set this path and we could save the configuration file, but I'm not sure if this would be the best solution on the Android world.